packages/seducer/src/reducer.js

```diff
@@ -0,0 +1,111 @@
+ *
+ * @return (state, actions) => (state, action) => {...}
+ *
+ * @param {unknown}                                state           The current state of the useReducer state can be anything
+ * @param {{key: {type:string, payload:unknown}}}  action          The action to be handle by the reducer and redirected to the their respective actions
+ *
+ * @return {unknown} nextState
+ */
+
+export const reducer = (actions, hasLogger, interceptors = {}) => (state, action) => {
+  const isAValidAction =
+    "type" in action && ([action.type] in actions || (interceptors && [action.type] in interceptors));
+
+  if (isAValidAction) {
+    const nextState = () => actions[action.type](state, action.payload);
+
+    if (interceptors && [action.type] in interceptors) {
+      return handleInterceptors({
+        action,
+        actions,
+        hasLogger,
+        interceptors,
+        nextState,
+        state,
+      });
+    }
+
+    return handleRegularRegularReducerFlow({
+      action,
+      hasLogger,
+      nextState,
+      state,
+    });
+  }
+
+  throw new Error(
+    `The type "${action.type}" lacks of an action implementation, did you forget to add it: 
+${action.type}(state, payload) { 
+  return state;
+} 
+
+to your action object`
+  );
+};
+
+/**
+ * Summary: Seducer permits to implement inversion of control without exposing the internal reducer and instead allowing the consumer to
+ *          extend its current action object, this has three benefits:
+ *
+ * First:   Abstract the concept of a reducer so the developer doesn't have to deal with
+ * Second:  Is quite simple to understand the concept of function and it's easy to explain to another developer that to override a function
+ *          has to declare another function with the same name, this is pretty common to do in OOP.
+ * Third:   In order to achieve inversion of control the library doesn't requires to export their internal reducer to the consumer
+ *          instead is transparent to them by simply overriding the already created actions.
+ *          In case duplication exists a third parameter will be sent to the action (state, payload, nextState) is
+ *          which is a function that the consumer can executed to generate the default nextState value that the component would generate
+ *          if wouldn't be intercepted.
+ *
+ *
+ * @param {{ type, payload }}                      action
+ * @param {{[key]:(state, payload)=> nextState}}   actions
+ * @param {boolean}                                hasLogger
+ * @param {{[key]:(state, payload)=> nextState}}   interceptors
+ * @param {unknown}                                nextState    A function that when its call generate the nextState for a giving action.
+ * @param {unknown}                                state
+ */
+
+function handleInterceptors({ action, actions, hasLogger, interceptors, nextState, state }) {
```

Comment:
Why don't we make `logger` an interceptor? The same way `redux` allows using logging middleware? That way you don't need to have an extra boolean option.

Reply:
`interceptors` are an advanced feature, is not need it in most cases. So forcing users to use it via the interceptors will be inconvenient for most of them. 

But I think would be cool as you suggest to have a logger property on the interceptors that you can take control of it. Will add that in the next version. 

I thought about adding a kind of middleware, but I don't know if that is outside of the scope of this small package. :)

Reply:
^ but in some sense the logger is already a middleware :P, let me refactor it for the next version and I will remove the boolean flags. :)